parsers/src/x3d/x3d_parser.cpp

```diff
@@ -75,7 +85,78 @@ void parse_tex_coord_index(std::string &&str, rcbe::geometry::Mesh::FacetStorage
         iss >> f.tex_coords_indices[0] >> f.tex_coords_indices[1] >> f.tex_coords_indices[2] >> sep;
     }
 }
+#ifdef R_TINYXML
+std::vector<rcbe::geometry::Mesh::TriangleType> parse_facets(const tinyxml2::XMLElement *element) {
+    auto *cind_attr = element->Attribute("coordIndex");
+    if (!cind_attr)
+        return {};
+
+    auto facets = parse_facets(std::string(cind_attr));
+
+    auto* tex_coords_elem = element->ToElement()->Attribute("texCoordIndex");
+    if (tex_coords_elem) {
+        parse_tex_coord_index(std::string(tex_coords_elem), facets);
+    }
+
+    return facets;
+}
+
+std::vector<rcbe::math::Vector3d> parse_vertices(const tinyxml2::XMLElement *element) {
+    auto* coord_elem = element->FirstChildElement("Coordinate");
+    if (!coord_elem)
+        return {};
+
+    auto* coord_point_elem = coord_elem->ToElement()->Attribute("point");
+    if (!coord_point_elem)
+        return {};
+
+    return parse_vertices(std::string(coord_point_elem));
+}
+
+std::vector<rcbe::math::Vector3d> parse_normals(const tinyxml2::XMLElement *element) {
+    auto* normal_elem = element->FirstChildElement("Normal");
+    if (!normal_elem)
+        return {};
+
+    auto* vector_elem = normal_elem->ToElement()->Attribute("vector");
+    if (!vector_elem)
+        return {};
+    
+    return parse_normals(std::string(vector_elem));
+}
+
+rcbe::geometry::Mesh parse_shape(const tinyxml2::XMLElement *shape_node) {
+    auto *ifs_elem = shape_node->FirstChildElement("IndexedFaceSet")->ToElement();
+    if (!ifs_elem)
+        return {};
+        
+    auto facets = parse_facets(ifs_elem);
+    auto vertices = parse_vertices(ifs_elem);
+    auto normals = parse_normals(ifs_elem);
```

Comment:
I'd suggest throwing if any of these is empty, or at least return a defaulted `Mesh`.

Reply:
There are no checks in original code as well, although I do believe that `boost:property_tree` throws, when we request a path which is not present in XML. So I'll throw here as well.

Reply:
Added exception in case if one of these is empty.